Sources/NIO/IntegerBitPacking.swift

```diff
@@ -0,0 +1,117 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+@usableFromInline
+enum _IntegerBitPacking {}
+
+extension _IntegerBitPacking {
+    @inlinable
+    static func packUU<Left: FixedWidthInteger & UnsignedInteger,
```

Comment:
- regarding the `typealias`: I think that'd be kind of counterproductive because the reader doesn't know what it means without expanding it (and I'd expect that they have some idea what the default ones are). But if others agree with you, I'm happy to change.
- regarding the argument labels: The order requirements aren't really there, it's typed so you can't pass the `medium` one to the `big` one by accident. On the _value level_, the only thing you need to do is match the order in `pack*` and `unpack*`. I first had it with labels but thought it's annoying especially if you do repeated applications. But again, if others think the labels _help_, then I'm happy to add them back.

Reply:
> * regarding the `typealias`: I think that'd be kind of counterproductive because the reader doesn't know what it means without expanding it (and I'd expect that they have some idea what the default ones are). But if others agree with you, I'm happy to change.

Fine by me, only a minor nit!

> * regarding the argument labels: The order requirements aren't really there, it's typed so you can't pass the `medium` one to the `big` one by accident. On the _value level_, the only thing you need to do is match the order in `pack*` and `unpack*`. I first had it with labels but thought it's annoying especially if you do repeated applications. But again, if others think the labels _help_, then I'm happy to add them back.

I do think it's clearer with the labels. And given this is underscored-private and only called from `IntegerBitPacking` I don't think having extra labels is too much of a burden?

```swift
func packU32U16U8(_ a: UInt32, _ b: UInt16, _ c: UInt8) -> UInt64 {
  // this is more obviously 'right' to me than 'packUUU(a, b, c)' 
  return _IntegerBitPacking.packUUU(big: a, medium: b, small: c)
}
```

Reply:
@glbrntt what makes `a` "big" and `b` "medium"? It's it weird if you did

```
packUUU(big: 0, medium: 1_000_000, small: .max)
```

?

Reply:
The relative width of their types make them "big" and "medium". I'm not saying that big and medium are the best names but they at least provide a hint that you can't pass the arguments in any order.

My concern is that as a caller it isn't clear at the point of use: there's nothing to differentiate the arguments of `packUUU` beyond the names of the generic types which only show up in code completion.

Reply:
I'll just throw some color into the bikeshed then, maybe then talk about width instead of size, e.g.:
`widest:, mid:, narrowest:`

Reply:
@glbrntt yes, but that refers to their _type_ not their _value_. They get packed left to right. I'm happy naming the types for example `Wide`/`Medium`/`Narrow` (similar to what @hassila suggests) and the labels maybe `left:middle:right:`. Would that work for you?

Reply:
Fine by me.

Reply:
done